### PR TITLE
Range-ify std.string.entab() by adding entabber()

### PR DESCRIPTION
**Pull request by WalterBright**
1. Added entabber() which takes a range as input and and returns a lazy range.
2. Rewrote entab() to use entabber().
3. Added support for nelSep.
4. Fixed bug where trailing spaces followed by newlines were not removed.
5. Added unittests to fill in missing cases.
6. 100% unittest coverage
7. Improved documentation.

### CODE REVIEW

File: std/string.d [hunk 1]
```diff
@@ -3171,108 +3171,269 @@ S detab(S)(S s, size_t tabSize = 8) @trusted pure
     Params:
         s       = String to convert.
         tabSize = Tab columns are $(D tabSize) spaces apart.
+
+    Returns:
+        GC allocated string with tabs replaced with spaces;
```
Comment (MartinNowak) [inline std/string.d]: `with spaces replaced with tabs`

File: std/string.d [hunk 2]
```diff
@@ -3171,108 +3171,269 @@ S detab(S)(S s, size_t tabSize = 8) @trusted pure
...
+        use $(LREF entabber) to not allocate.
+
+    See_Also:
+        $(LREF entabber)
  +/
-S entab(S)(S s, size_t tabSize = 8) @trusted pure
+S entab(S)(S s, size_t tabSize = 8) @trusted
```
Comment (MartinNowak) [inline std/string.d]: Should still be pure, right?

File: std/string.d [hunk 3]
```diff
@@ -3171,108 +3171,269 @@ S detab(S)(S s, size_t tabSize = 8) @trusted pure
...
-    import std.exception : assumeUnique;
+    import std.array;
+    return cast(S)(entabber(s, tabSize).array);
+}
 
-    bool changes = false;
-    alias C = Unqual!(typeof(s[0]));
-    C[] result;
+///
+unittest
```
Comment (MartinNowak) [inline std/string.d]: Better add `@safe pure` and whatnot to the unittest.

File: std/string.d [hunk 4]
```diff
@@ -3171,108 +3171,269 @@ S detab(S)(S s, size_t tabSize = 8) @trusted pure
...
 
-    foreach (size_t i, dchar c; s)
+    Returns:
+        lazy forward range with spaces replaced with tabs
+
+    See_Also:
+        $(LREF entab)
+  +/
+auto entabber(Range)(Range r, size_t tabSize = 8)
+    if (isForwardRange!Range)
```
Comment (MartinNowak) [inline std/string.d]: Why aren't input ranges supported?

File: std/string.d [hunk 5]
```diff
@@ -3171,108 +3171,269 @@ S detab(S)(S s, size_t tabSize = 8) @trusted pure
...
-            {   nwhite++;
-                nspaces++;
+                index = 1;
+                dc = c;
+                if (c == ' ' || c == '\t')
+                {
+                    // Consume input until a non-blank is encountered
+                    size_t startcol = column;
+                    C cx;
+                    static if (isSomeString!Range)
```
Comment (MartinNowak) [inline std/string.d]: Wouldn't it be better to work on [`_input.representation`](http://dlang.org/library/std/string/representation.html) rather than duplicating all this code for narrow strings?

File: std/string.d [hunk 6]
```diff
@@ -3171,108 +3171,269 @@ S detab(S)(S s, size_t tabSize = 8) @trusted pure
...
-        void change()
+        @property C getFront()
         {
-            changes = true;
-            result = null;
-            result.length = s.length;
-            result.length = i;
-            result[0 .. i] = s[0 .. i];
+            static if (isSomeString!Range)
+                return _input[0];       // avoid autodecode
```
Comment (MartinNowak) [inline std/string.d]: That's problematic because, because you're disabling decoding for the user of entabber, but still present a dchar range.

**Comment by MartinNowak** [comment]
I think it would be helpful if you didn't mix your GC changes with a crusade against autodecode.
At many places you could probably avoid duplicating code by using [std.string.representation](http://dlang.org/library/std/string/representation.html).
You'll have to decode anyhow, because a dchar range can't return UTF-8 codeunits and a char range isn't usable as string (e.g. writeln).

File: std/string.d (hunk 2 above)
Comment (WalterBright) [inline std/string.d]: It's supposed to infer its attributes, being a template.

File: std/string.d (hunk 3 above)
Comment (WalterBright) [inline std/string.d]: That's purely a documentation unittest. The real ones are later and are so marked.

File: std/string.d (hunk 1 above)
Comment (WalterBright) [inline std/string.d]: Copy-pasta strikes again!

File: std/string.d (hunk 4 above)
Comment (WalterBright) [inline std/string.d]: Need lookahead.

File: std/string.d (hunk 5 above)
Comment (WalterBright) [inline std/string.d]: Doing it this way is a bit faster because the range path requires using `.save`.

**Comment by WalterBright** [comment]
> I think it would be helpful if you didn't mix your GC changes with a crusade against autodecode.

The original did not autodecode, either, it returned a string in the same encoding as the input.

> At many places you could probably avoid duplicating code by using std.string.representation.

byCodeUnit would be better, but I prefer a separate path because the arrays don't require use of .save.

> You'll have to decode anyhow, because a dchar range can't return UTF-8 codeunits

entabber is designed to return the same char type as its input. A dchar range would return a dchar range. If you want to convert the output to UTF-8, tack on .byChar.

> and a char range isn't usable as string (e.g. writeln).

That would be a bug in writeln, and is not part of the charter of entabber.

File: std/string.d (hunk 6 above)
Comment (WalterBright) [inline std/string.d]: No, it returns an instance of C, which is the ElementEncodingType

File: std/string.d [hunk 7]
```diff
@@ -3171,108 +3171,269 @@ S detab(S)(S s, size_t tabSize = 8) @trusted pure
...
-
-                ptrdiff_t j = result.length - nspaces;
-                auto ntabs = (((column - nspaces) % tabSize) + nspaces) / tabSize;
-                result.length = j + ntabs;
-                result[j .. j + ntabs] = '\t';
-                nwhite += ntabs - nspaces;
-                nspaces = 0;
+                foreach (c; _input)
+                {
+                    if (c != ' ' && c != '\t')
```
Comment (Panke) [inline std/string.d]: Are these the only spaces we should be looking for? What about http://en.wikipedia.org/wiki/Whitespace_character without the line breaking characters?

**Comment by Panke** [comment]
LGTM

File: std/string.d (hunk 6 above)
Comment (MartinNowak) [inline std/string.d]: Ah OK, misread that as ElementType.

File: std/string.d (hunk 7 above)
Comment (MartinNowak) [inline std/string.d]: Fixing entab is not part of this pull request, and it's questionable whether you want to convert a no break space into a tab.

**Comment by MartinNowak** [comment]
Let me repeat an important point though. You're introducing a new type of narrow strings here, but at this point, almost no function in phobos can deal with a range of chars. This makes entabber unusable without byDchar, and everyone trying to use it would have to find that out by itself.

**Comment by MartinNowak** [comment]
So either we autodecode here like in any other range algorithm, or we change all the rest of phobos to handle such ranges (e.g. find), or we tell people to use byDchar.

**Comment by WalterBright** [comment]
I know that other algorithms, like std.algorithm.equal, currently fail if presented with a range of chars or wchars. The only solutions are:

1. fix the algorithms
2. not have ranges of chars or wchars

I choose option (1). (2) would severely compromise Phobos' utility.
